README.md
## Blogpost - Implementing time-critical "Cloud-To-Device" IoT message patterns on AWS IoT Core

This repository contains the artifcat related to the Implementing time-critical "Cloud-To-Device" IoT message patterns on AWS IoT Core blogpost.

## Security

See [CONTRIBUTING](CONTRIBUTING.md#security-issue-notifications) for more information.

## License

This library is licensed under the MIT-0 License. See the LICENSE file.

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

This project follows [Keep a Changelog](https://keepachangelog.com/en/1.0.0/) format for changes and adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.0] - 2022-08-17

### Fixed

* N/A

### Added

* First release of IoT Direct Invocations Lambda and sample client

sam/test-client/index.py
##########################################################################################
#
#  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
#  SPDX-License-Identifier: MIT-0
#
#  Permission is hereby granted, free of charge, to any person obtaining a copy of this
#  software and associated documentation files (the "Software"), to deal in the Software
#  without restriction, including without limitation the rights to use, copy, modify,
#  merge, publish, distribute, sublicense, and/or sell copies of the Software, and to
#  permit persons to whom the Software is furnished to do so.
#
#  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
#  INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
#  PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
#  HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
#  OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
#  SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
#
##########################################################################################

import argparse
import json
import uuid
from awscrt import io, mqtt
from awsiot import mqtt_connection_builder
import time
from datetime import datetime
import random

g_received_count = 0
g_is_command_received = False
g_message_id = None
g_request = None
g_method = None
g_caller = None
g_timeout = 10000

g_operations = ["add", "sum", "div", "mul"]

def get_milliseconds():
    return round(time.time() * 1000)

# Callback when the subscribed topic receives a message
def on_message_received(topic, payload, dup, qos, retain, **kwargs):
    print(f"Received message from topic '{topic}': {payload}")

    global g_is_command_received
    global g_message_id
    global g_request
    global g_method
    global g_caller

    topicsplit = topic.split("/")
    g_method = topicsplit[1]
    g_caller = topicsplit[2]

    obj = json.loads(payload)

    print ("Obj: {obj}")

    g_is_command_received = False
    if ('id' in obj.keys()):
        g_is_command_received = True
        g_message_id = obj['id']

    if ('request' in obj.keys()):
        g_request = obj['request']

    if ('operation' in obj.keys()):
        req_op = obj['operation']
        if (req_op.upper() in (op.upper() for op in g_operations)):
            print ("Allowed operator: {req_op}")
    else:
        print ("Missing parameter 'operation'")



def connect(device_id, endpoint, port, cert_path, key_path):
    ts_start = get_milliseconds()

    event_loop_group = io.EventLoopGroup(1)
    host_resolver = io.DefaultHostResolver(event_loop_group)
    client_bootstrap = io.ClientBootstrap(event_loop_group, host_resolver)

    ################################################################################
    ### Create & Connect Client

    mqtt_connection = mqtt_connection_builder.mtls_from_path(
        client_id=device_id,
        endpoint=endpoint,
        port=int(port),
        cert_filepath=cert_path,
        pri_key_filepath=key_path,
        client_bootstrap=client_bootstrap,
        clean_session=False,
        keep_alive_secs=30
    )

    connect_future = mqtt_connection.connect()
    connect_future.result()

    print("Connected!")

    ################################################################################
    ### Subscribe to Topic

    message_topic = f"{device_id}/#"

    print(f"Subscribing to topic '{message_topic}' ...")

    subscribe_future, packet_id = mqtt_connection.subscribe(
        topic=message_topic,
        qos=mqtt.QoS.AT_LEAST_ONCE,
        callback=on_message_received)

    subscribe_result = subscribe_future.result()

    print("Subscribed with {}".format(str(subscribe_result['qos'])))

    ################################################################################
    ### Wait for Command

    global g_is_command_received

    print("Waiting for command ...")

    while True:

        if g_is_command_received:
            res_time = get_milliseconds()

            print(f"Send ACK")
            g_is_command_received = False

            message_json = json.dumps({
                    'id':g_message_id,
                    'response':f"Your request was '{g_request}'. Some random response: '{random.random() * 1000}'"
                })

            g_message_ack_topic = f"{g_caller}/{g_message_id}/{device_id}/{g_method}/ack"

            print(f"Publishing message '{message_json}' on topic '{g_message_ack_topic}'")

            res, _ = mqtt_connection.publish(
                topic=g_message_ack_topic,
                payload=message_json,
                qos=mqtt.QoS.AT_LEAST_ONCE
            )

            res.result()

            print(f"Response '{res}' elapsed: {get_milliseconds() - res_time} ms")

        time.sleep(0.001)

def parse_args():
    parser = argparse.ArgumentParser(description="Device Shadow sample keeps a property in sync across client and server")

    parser.add_argument('--endpoint', required=True, help="Your AWS IoT custom endpoint, not including a port. Ex: \"w6zbse3vjd5b4p-ats.iot.us-west-2.amazonaws.com\"")
    parser.add_argument('--port', default="8883", help="Endpoint port")
    parser.add_argument('--cert', required=True, help="File path to your client certificate, in PEM format")
    parser.add_argument('--key', required=True, help="File path to your private key file, in PEM format")
    parser.add_argument('--client-id', default="test-" + str(uuid.uuid4()), help="Client ID for MQTT connection.")

    return parser.parse_args()

if __name__ == '__main__':
    g_args = parse_args()

    print(f"args: {g_args}")

    print(f"Endpoint: {g_args.endpoint}")
    print(f"Port: {g_args.port}")
    print(f"Cert Path: {g_args.cert}")
    print(f"Key Path: {g_args.key}")
    print(f"Client ID: {g_args.client_id}")

    connect(g_args.client_id, g_args.endpoint, g_args.port, g_args.cert, g_args.key)


sam/iot-lambda-client/endpoint/app.py
##########################################################################################
#
#  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
#  SPDX-License-Identifier: MIT-0
#
#  Permission is hereby granted, free of charge, to any person obtaining a copy of this
#  software and associated documentation files (the “Software”), to deal in the Software
#  without restriction, including without limitation the rights to use, copy, modify,
#  merge, publish, distribute, sublicense, and/or sell copies of the Software, and to
#  permit persons to whom the Software is furnished to do so.
#
#  THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
#  INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
#  PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
#  HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
#  OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
#  SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
#
##########################################################################################

from asyncio import events
import json
import uuid
from awscrt import io, mqtt, auth, http
from awsiot import mqtt_connection_builder
import time
from datetime import datetime
import os
import sys
import threading

g_client_id_prefix = ""
g_endpoint = ""

g_message_ack_topic_prefix = ""
g_message_topic_prefix = ""

g_received_count = 0

g_is_ack_received = False
g_message_id = None
g_device_response = None

g_timeout = 10


def get_milliseconds():
    return round(time.time() * 1000)

# Callback when the subscribed topic receives a message
def on_message_received(topic, payload, dup, qos, retain, **kwargs):
    print(f"Received message from topic '{topic}': {payload}")

    global g_is_ack_received
    global g_message_id
    global g_device_response

    obj = json.loads(payload)

    g_is_ack_received = False
    g_device_response = None

    if ('response' in obj.keys()) and (not obj['response'] == None):
        g_device_response = obj['response']

    # Ack as last thing for concurrency
    if ('id' in obj.keys()) and (obj['id'] == g_message_id):
        g_is_ack_received = True

    returnobj = {
            'g_device_response': g_device_response,
            'g_is_ack_received': g_is_ack_received
        }

    return returnobj


def do_request(p_method, p_request, p_timeout):
    ts_start = get_milliseconds()

    global g_is_ack_received
    global g_endpoint
    global g_client_id_prefix

    g_is_ack_received = False

    print(
        f"Performing request '{p_request}' for method '{p_method}' with timeout '{p_timeout}'")

    event_loop_group = io.EventLoopGroup(1)
    host_resolver = io.DefaultHostResolver(event_loop_group)
    client_bootstrap = io.ClientBootstrap(event_loop_group, host_resolver)
    credentials_provider = auth.AwsCredentialsProvider.new_default_chain(client_bootstrap)
    proxy_options = None

    ################################################################################
    #  Create & Connect Client

    mqtt_client_id = f"{g_client_id_prefix}_{g_message_id}"

    print(f"Client ID: '{mqtt_client_id}'")

    mqtt_connection = mqtt_connection_builder.websockets_with_default_aws_signing(
                endpoint=g_endpoint,
                client_bootstrap=client_bootstrap,
                region=g_region,
                credentials_provider=credentials_provider,
                http_proxy_options=proxy_options,
                client_id=mqtt_client_id,
                clean_session=False,
                keep_alive_secs=30)

    connect_future = mqtt_connection.connect()
    connect_future.result()

    print("Connected!")

    ################################################################################
    # Subscribe to Topic

    message_ack_topic = f"{g_message_ack_topic_prefix}"

    print(f"Subscribing to topic '{message_ack_topic}'...")
    subscribe_future, packet_id = mqtt_connection.subscribe(
        topic=message_ack_topic,
        qos=mqtt.QoS.AT_LEAST_ONCE,
        callback=on_message_received)

    subscribe_result = subscribe_future.result()
    print("Subscribed with {}".format(str(subscribe_result['qos'])))

    ################################################################################
    # Send Command

    message = {
        "timestamp": str(get_milliseconds()),
        "request": p_request,
        "id": g_message_id
    }

    message_json = json.dumps(message)

    message_topic = f"{g_message_topic_prefix}"

    print(f"Publishing message '{message_json}' on topic '{message_topic}'")

    mqtt_connection.publish(
        topic=message_topic,
        payload=message_json,
        qos=mqtt.QoS.AT_MOST_ONCE)

    ################################################################################
    # Wait for ACK

    wtstart = get_milliseconds()
    is_timed_out = False
    while not (g_is_ack_received or is_timed_out):
        is_timed_out = (get_milliseconds() - wtstart) > (p_timeout * 1000)
        time.sleep(0.001)
    print(
        f"Waiting time: {get_milliseconds() - wtstart} ms, timeout: {p_timeout * 1000} ms")

    ################################################################################
    # Evaluate Response OR Timeout

    response = None

    print(f"Is Timed-Out: {is_timed_out}")
    print(f"Is ACK Received: {g_is_ack_received}")
    print(f"Device Response: {g_device_response}")

    if is_timed_out:
        response = {
            'statusCode': 500,
            'body': json.dumps({
                'result': 'timeout',
                'elapsed': str(get_milliseconds() - ts_start),
            }),
            'headers': {
                'Content-Type': 'application/json'
            }
        }
    elif g_is_ack_received:
        response = {
            'statusCode': 200,
            'body': json.dumps({
                'result': 'ok',
                'elapsed': str(get_milliseconds() - ts_start),
                'response': g_device_response
            }),
            'headers': {
                'Content-Type': 'application/json'
            }
        }
    else:
        response = {
            'statusCode': 500,
            'body': json.dumps({
                'result': 'internal_error',
                'elapsed': str(get_milliseconds() - ts_start),
            }),
            'headers': {
                'Content-Type': 'application/json'
            }
        }

    ################################################################################
    # Disconnect Client

    print("Disconnecting...")
    disconnect_future = mqtt_connection.disconnect()
    disconnect_future.result()
    print("Disconnected!")

    print(f"Process Elapsed: {get_milliseconds() - ts_start}")

    return response


def handler(event, context):

    global g_client_id_prefix
    global g_endpoint
    global g_message_ack_topic_prefix
    global g_message_topic_prefix
    global g_region
    
    g_client_id_prefix = os.environ.get("CLIENT_ID_PREFIX")
    g_endpoint = os.environ.get("WSS_ENDPOINT")
    g_region = os.environ.get("AWS_REGION")

    method = None
    request = None
    target = None
    timeout = None
    response = None

    print (f"Events: '{event}'")

    if ('queryStringParameters' in event.keys()):
        params = event['queryStringParameters']

        print (f"Params: '{params}'")

        if ('method' in params.keys()):
            method = params['method']

        if ('request' in params.keys()):
            request = params['request']

        if ('target' in params.keys()):
            target = params['target']

        if ('timeout' in params.keys()):
            try:
                param_timeout = params['timeout']
                timeout = int(param_timeout)
                if timeout <= 0 or timeout > 60:
                    print(
                        f"Timeout value value '{timeout}' is out of range (0-60), replace with default one")
                    timeout = g_timeout
            except:
                print(
                    f"Timeout value '{param_timeout}' is not a number, getting default one '{g_timeout}'.")
                timeout = g_timeout
        else:
            print("No timeout value passed, assuming default one")
            timeout = g_timeout

    if method != None and request != None and target != None:

        global g_message_id
        g_message_id = str(uuid.uuid4())

        print(f"Message ID: {g_message_id}")

        g_message_ack_topic_prefix = f"{g_client_id_prefix}/{g_message_id}/{target}/{method}/ack"
        g_message_topic_prefix     = f"{target}/{method}/{g_client_id_prefix}/{g_message_id}"

        print(f"Message topic: {g_message_topic_prefix}")
        print(f"ACK topic: {g_message_ack_topic_prefix}")

        response = do_request(method, request, timeout)
    else:
        response = {
            'statusCode': 500,
            'body': json.dumps({
                'result': 'invalid_parameters',
                'response': 'The \'method\' and/or \'request\' and/or \'target\' query parameters are missing!'
            }),
            'headers': {
                'Content-Type': 'application/json'
            }
        }

    print(f"Response: {response}")

    return response